File: packages/go/pkg/components/keyboard/reply.go
package keyboard

import "github.com/Nam088/telebot/packages/go/pkg/types"

// KeyboardButton represents one button of a custom reply keyboard.
//
// If none of the optional request flags are set, the button text is sent to
// the bot as a plain message when the button is pressed. At most one field
// other than Text, IconCustomEmojiID and Style may be set.
//
// Telegram API: https://core.telegram.org/bots/api#keyboardbutton
type KeyboardButton struct {
	// Text is the label shown on the button.
	Text string `json:"text"`
	// IconCustomEmojiID is the unique identifier of the custom emoji shown
	// before the button text.
	IconCustomEmojiID string `json:"icon_custom_emoji_id,omitempty"`
	// Style is the visual style of the button: "danger", "success" or
	// "primary". When empty, an app-specific style is used.
	Style string `json:"style,omitempty"`
	// RequestUsers, when set, opens a list of suitable users whose identifiers
	// are sent to the bot in a "users_shared" service message. Available in
	// private chats only.
	RequestUsers *types.KeyboardButtonRequestUsers `json:"request_users,omitempty"`
	// RequestChat, when set, opens a list of suitable chats whose identifier is
	// sent to the bot in a "chat_shared" service message. Available in private
	// chats only.
	RequestChat *types.KeyboardButtonRequestChat `json:"request_chat,omitempty"`
	// RequestManagedBot, when set, asks the user to create and share a bot
	// managed by the current bot. Available in private chats only to bots that
	// enabled bot management in the @BotFather Mini App.
	RequestManagedBot *types.KeyboardButtonRequestManagedBot `json:"request_managed_bot,omitempty"`
	// RequestContact, when true, sends the user's phone number as a contact
	// when the button is pressed. Available in private chats only.
	RequestContact bool `json:"request_contact,omitempty"`
	// RequestLocation, when true, sends the user's current location when the
	// button is pressed. Available in private chats only.
	RequestLocation bool `json:"request_location,omitempty"`
	// RequestPoll, when set, asks the user to create a poll and send it to the
	// bot when the button is pressed. Available in private chats only.
	RequestPoll *types.KeyboardButtonPollType `json:"request_poll,omitempty"`
	// WebApp, when set, launches the described Web App when the button is
	// pressed. Available in private chats only.
	WebApp *types.WebAppInfo `json:"web_app,omitempty"`
}

// ReplyKeyboardMarkup represents a custom keyboard with reply options that is
// displayed in place of the user's on-screen keyboard. Not supported in
// channels and for messages sent on behalf of a business account.
//
// Telegram API: https://core.telegram.org/bots/api#replykeyboardmarkup
type ReplyKeyboardMarkup struct {
	// Keyboard is the array of button rows, each represented as a slice of
	// KeyboardButton objects.
	Keyboard [][]KeyboardButton `json:"keyboard"`
	// IsPersistent requests clients to always show the keyboard when the
	// regular keyboard is hidden.
	IsPersistent bool `json:"is_persistent,omitempty"`
	// ResizeKeyboard requests clients to resize the keyboard vertically for
	// optimal fit instead of just making it smaller.
	ResizeKeyboard bool `json:"resize_keyboard,omitempty"`
	// OneTimeKeyboard requests clients to hide the keyboard as soon as it has
	// been used.
	OneTimeKeyboard bool `json:"one_time_keyboard,omitempty"`
	// InputFieldPlaceholder is the placeholder shown in the input field when
	// the keyboard is active; 1-64 characters.
	InputFieldPlaceholder string `json:"input_field_placeholder,omitempty"`
	// Selective limits the keyboard to users @mentioned in the message and to
	// the sender of the replied-to message.
	Selective bool `json:"selective,omitempty"`
	// ForceReply requests clients to show the reply interface to the user, as
	// if they had selected the bot's message and tapped "Reply".
	ForceReply bool `json:"force_reply,omitempty"`
}

// ReplyKeyboardRemove requests clients to remove the current custom keyboard
// and display the default letter-keyboard.
//
// Telegram API: https://core.telegram.org/bots/api#replykeyboardremove
type ReplyKeyboardRemove struct {
	// RemoveKeyboard must be true; the user will not be able to summon this
	// keyboard again.
	RemoveKeyboard bool `json:"remove_keyboard"`
	// Selective limits the removal to users @mentioned in the message and to
	// the sender of the replied-to message.
	Selective bool `json:"selective,omitempty"`
}

// ForceReply requests clients to display a reply interface to the user, as if
// they had selected the bot's message and tapped "Reply".
//
// Telegram API: https://core.telegram.org/bots/api#forcereply
type ForceReply struct {
	// ForceReply must be true to show the reply interface.
	ForceReply bool `json:"force_reply"`
	// InputFieldPlaceholder is the placeholder shown in the input field when
	// the reply is active; 1-64 characters.
	InputFieldPlaceholder string `json:"input_field_placeholder,omitempty"`
	// Selective limits the reply interface to users @mentioned in the message
	// and to the sender of the replied-to message.
	Selective bool `json:"selective,omitempty"`
}

// ReplyKeyboardOption mutates a ReplyKeyboard builder to configure how the
// resulting markup is presented to the client.
type ReplyKeyboardOption func(*ReplyKeyboard)

// WithResizeKeyboard requests clients to resize the keyboard vertically for
// optimal fit.
func WithResizeKeyboard() ReplyKeyboardOption {
	return func(k *ReplyKeyboard) {
		k.resizeKeyboard = true
	}
}

// WithOneTimeKeyboard requests clients to hide the keyboard as soon as it has
// been used.
func WithOneTimeKeyboard() ReplyKeyboardOption {
	return func(k *ReplyKeyboard) {
		k.oneTimeKeyboard = true
	}
}

// WithInputFieldPlaceholder sets the placeholder shown in the input field when
// the keyboard is active; 1-64 characters.
func WithInputFieldPlaceholder(placeholder string) ReplyKeyboardOption {
	return func(k *ReplyKeyboard) {
		k.inputFieldPlaceholder = placeholder
	}
}

// ReplyKeyboard is a fluent builder for constructing ReplyKeyboardMarkup values.
type ReplyKeyboard struct {
	rows                  [][]KeyboardButton
	resizeKeyboard        bool
	oneTimeKeyboard       bool
	inputFieldPlaceholder string
}

// NewReplyKeyboard creates a new empty reply keyboard builder.
//
// Example:
//
//	kb := keyboard.NewReplyKeyboard(
//	    keyboard.WithResizeKeyboard(),
//	    keyboard.WithOneTimeKeyboard(),
//	    keyboard.WithInputFieldPlaceholder("Pick one..."),
//	).
//	    AddButton("Option A").
//	    AddButton("Option B").
//	    AddRow().
//	    AddButton("Help").
//	    Build()
func NewReplyKeyboard(opts ...ReplyKeyboardOption) *ReplyKeyboard {
	k := &ReplyKeyboard{
		rows: make([][]KeyboardButton, 0),
	}
	for _, opt := range opts {
		opt(k)
	}
	return k
}

// AddRow starts a new row of buttons. Subsequent AddButton calls append to this
// row. Pre-built buttons may be passed to populate the row immediately.
func (k *ReplyKeyboard) AddRow(buttons ...KeyboardButton) *ReplyKeyboard {
	if len(buttons) > 0 {
		k.rows = append(k.rows, buttons)
	} else {
		k.rows = append(k.rows, []KeyboardButton{})
	}
	return k
}

// AddButton appends a plain text button to the current row. If no row exists
// yet, a first row is created implicitly.
//
// Parameters:
//   - text: Label shown on the button, sent to the bot as a message when pressed.
func (k *ReplyKeyboard) AddButton(text string) *ReplyKeyboard {
	btn := KeyboardButton{Text: text}
	if len(k.rows) == 0 {
		k.rows = append(k.rows, []KeyboardButton{btn})
	} else {
		lastIdx := len(k.rows) - 1
		k.rows[lastIdx] = append(k.rows[lastIdx], btn)
	}
	return k
}

// Build compiles the configured layout into a *ReplyKeyboardMarkup. Empty rows
// are dropped from the final keyboard.
func (k *ReplyKeyboard) Build() *ReplyKeyboardMarkup {
	rows := make([][]KeyboardButton, 0, len(k.rows))
	for _, row := range k.rows {
		if len(row) > 0 {
			rows = append(rows, row)
		}
	}
	return &ReplyKeyboardMarkup{
		Keyboard:              rows,
		ResizeKeyboard:        k.resizeKeyboard,
		OneTimeKeyboard:       k.oneTimeKeyboard,
		InputFieldPlaceholder: k.inputFieldPlaceholder,
	}
}

File: packages/node/src/components/keyboard/reply.ts
/**
 * Reply keyboard markup and fluent builder.
 *
 * @packageDocumentation
 */

import type { ReplyKeyboardMarkup, KeyboardButton } from "../../client/types.js";

/**
 * Configuration options for {@link ReplyKeyboard}.
 */
export interface ReplyKeyboardOptions {
  /**
   * Requests clients to resize the keyboard vertically for optimal fit.
   *
   * @defaultValue `false`
   */
  resize_keyboard?: boolean;
  /**
   * Requests clients to hide the keyboard as soon as it's been used.
   *
   * @defaultValue `false`
   */
  one_time_keyboard?: boolean;
  /**
   * Requests clients to always show the keyboard when the regular keyboard is hidden.
   *
   * @defaultValue `false`
   */
  is_persistent?: boolean;
  /**
   * Placeholder to be shown in the input field when the keyboard is active.
   */
  input_field_placeholder?: string;
  /**
   * Use this parameter if you want to show the keyboard to specific users only.
   */
  selective?: boolean;
}

/**
 * Fluent builder for creating {@link ReplyKeyboardMarkup} objects.
 *
 * @example
 * ```ts
 * const keyboard = new ReplyKeyboard({ resize_keyboard: true })
 *   .text("Option A")
 *   .text("Option B")
 *   .row()
 *   .requestLocation("Share Location");
 *
 * await bot.sendMessage({
 *   chat_id: 12345,
 *   text: "Make a selection:",
 *   reply_markup: keyboard.build(),
 * });
 * ```
 */
export class ReplyKeyboard {
  private readonly rows: KeyboardButton[][] = [[]];
  private readonly options: ReplyKeyboardOptions;

  /**
   * Constructs a new {@link ReplyKeyboard} builder instance.
   *
   * @param options - Optional settings for reply keyboard presentation.
   */
  constructor(options: ReplyKeyboardOptions = {}) {
    this.options = options;
  }

  /**
   * Appends a plain text button to the current row.
   *
   * @param text - Text sent to the bot as a message when the button is pressed.
   * @returns This {@link ReplyKeyboard} instance for chaining.
   */
  public text(text: string): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({ text });
    }
    return this;
  }

  /**
   * Appends a button that prompts the user to share their contact.
   *
   * @param text - Button label text.
   * @returns This {@link ReplyKeyboard} instance for chaining.
   */
  public requestContact(text: string): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({ text, request_contact: true });
    }
    return this;
  }

  /**
   * Appends a button that prompts the user to share their location.
   *
   * @param text - Button label text.
   * @returns This {@link ReplyKeyboard} instance for chaining.
   */
  public requestLocation(text: string): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({ text, request_location: true });
    }
    return this;
  }

  /**
   * Appends a button that prompts the user to create and send a poll.
   *
   * @param text - Button label text.
   * @param type - Optional poll type restriction (`"quiz"` or `"regular"`).
   * @returns This {@link ReplyKeyboard} instance for chaining.
   */
  public requestPoll(text: string, type?: "quiz" | "regular"): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({ text, request_poll: { type } });
    }
    return this;
  }

  /**
   * Appends a button that launches a Telegram Mini App Web App.
   *
   * @param text - Button label text.
   * @param url - HTTPS URL of the Web App.
   * @returns This {@link ReplyKeyboard} instance for chaining.
   */
  public webApp(text: string, url: string): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({ text, web_app: { url } });
    }
    return this;
  }

  /**
   * Appends a button that prompts the user to select one or more users.
   *
   * @param text - Button label text.
   * @param requestId - Signed 32-bit identifier of the request.
   * @param options - Additional user selection filters.
   * @returns This {@link ReplyKeyboard} instance for chaining.
   */
  public requestUsers(
    text: string,
    requestId: number,
    options: {
      user_is_bot?: boolean;
      user_is_premium?: boolean;
      max_quantity?: number;
      request_name?: boolean;
      request_username?: boolean;
      request_photo?: boolean;
    } = {},
  ): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({
        text,
        request_users: {
          request_id: requestId,
          ...options,
        },
      });
    }
    return this;
  }

  /**
   * Appends a button that prompts the user to select a chat.
   *
   * @param text - Button label text.
   * @param requestId - Signed 32-bit identifier of the request.
   * @param chatIsChannel - Pass `true` to request a channel chat, `false` for a group.
   * @param options - Additional chat selection criteria.
   * @returns This {@link ReplyKeyboard} instance for chaining.
   */
  public requestChat(
    text: string,
    requestId: number,
    chatIsChannel: boolean,
    options: {
      chat_is_forum?: boolean;
      chat_has_username?: boolean;
      chat_is_created?: boolean;
      bot_is_member?: boolean;
      request_title?: boolean;
      request_username?: boolean;
      request_photo?: boolean;
    } = {},
  ): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({
        text,
        request_chat: {
          request_id: requestId,
          chat_is_channel: chatIsChannel,
          ...options,
        },
      });
    }
    return this;
  }

  /**
   * Advances the builder to start a new keyboard row.
   *
   * @returns This {@link ReplyKeyboard} instance for chaining.
   */
  public row(): this {
    if (
      this.rows.length === 0 ||
      (this.rows[this.rows.length - 1] && this.rows[this.rows.length - 1]!.length > 0)
    ) {
      this.rows.push([]);
    }
    return this;
  }

  /**
   * Builds and returns the final {@link ReplyKeyboardMarkup} structure.
   *
   * @returns The constructed {@link ReplyKeyboardMarkup}.
   */
  public build(): ReplyKeyboardMarkup {
    const keyboard = this.rows.filter((row) => row.length > 0);
    return {
      keyboard,
      resize_keyboard: this.options.resize_keyboard,
      one_time_keyboard: this.options.one_time_keyboard,
      is_persistent: this.options.is_persistent,
      input_field_placeholder: this.options.input_field_placeholder,
      selective: this.options.selective,
    };
  }
}
